supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala

```diff
@@ -470,7 +476,16 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
   }
 
   override def mapOutputWomFile(womFile: WomFile): WomFile = {
-    womFileToGcsPath(generateJesOutputs(jobDescriptor))(womFile)
+    val cloudOutputWomFile: WomFile = womFileToGcsPath(generateJesOutputs(jobDescriptor))(womFile)
+    val cloudOutputPath: Path = backendEngineFunctions.buildPath(cloudOutputWomFile.value)
+
+    // It's not really exceptional for an optional output file to be missing but this is how that fact is communicated
+    // to the caller.
+    // .exists() on the file itself is fast but not 100% compatible with the gs:// URLs supported by gsutil.
+    // AsyncIo appears to be more compatible but a lot slower when used synchronously for individual files like this,
+    // so only use that if exists() returns false.
+    if (!cloudOutputPath.exists() && !asyncIo.exists(cloudOutputPath)) throw new FileNotFoundException(s"Could not process output, file not found: ${cloudOutputPath.pathAsString}")
```

Comment:
I'm surprised that it takes +1h with the asyncIo. I'm wondering if it's not the await result that makes it so much slower rather than the fact that it's going through the Io actor (which makes it slower but shouldn't be that much). My main concern is that the `.exists` on the path calls directly NIO which amongst other things doesn't have our custom retries which prevent known transient failures. I'd be curious to see what a Futurized version (without Await.result) would do performance wise, but I imagine that it would be extremely annoying to Futurize all this so... MaybeLaterTM

Reply:
I think you explained it to me already but can you re-explain why we need here to check that the files indeed do exist ? Doesn't the delocalization fail if they don't ?

Reply:
This follows the behavior of the SFS backend which checks if an output file exists and if it does not throws a FNFE. From the context of evaluating an optional File output this gets turned into a None, which then shows up as a null-valued output in the metadata JSON. This conformance test does not expect to see the nonexistent optional output file in the output JSON, which I guess the OutputManipulator is helpful enough to remove for null values? This existence check was also helpful in pointing out that the `read_tsv` Centaur test doesn't actually work on PAPI since the files in the generated TSV are never delocalized from the VM.

Reply:
I'll kick off another build with just async IO, maybe yesterday was just a bad cloud day.

Reply:
Oh ok I see what you're saying with the `read_tsv` test. I just think it's potentially expensive to add a request for every job output file (I'm not saying it's bad but we should be considerate of the performance impact, especially when we see how much Cromwell struggles these days).
If anything IMO the test is not thorough enough in the first place and should do a `read_string` or try to access the files in the array to prove that they are indeed delocalized.
If we can get away with relying on the PAPI delocalization to determine file existence without having to double check all of them in Cromwell afterwards it would be a big win.
But I don't think I understand all the intricacies with optional outputs so if you think it's necessary I trust your judgment 😄

Reply:
Happy to talk this over more because tbh I'm a little apprehensive about the performance implications of adding this check as well, I just don't know another way to do it atm. 🙂 I'm not sure what you mean by having the delocalization determine file existence (outside the context of the VM)?

Reply:
The case of the `read_tsv` test is actually similar to a `cwl.output.json` containing file paths. The file paths in the tsv as well as the ones in the `cwl.output.json` are what we could call "dynamic outputs" in that they're unknown when the job starts. They have to be "discovered" once the job is complete and delocalized appropriately. I guess what I'm saying is that for non-dynamic outputs (the majority of them), we'll have an explicit delocalization action (as they're known when we create the job). For those, we can rely on the fact that `gsutil cp` will fail if the file is not there after the command runs (unless it's optional). If we assume that a successful `gsutil cp` guarantees that the file is in GCS, we shouldn't need to test their existence in the output evaluation.
For dynamic outputs, it's true that if we don't do anything special they will not be delocalized. However, provided that they are actually used in later tasks (or accessed in any way via `read_string` or whatever), the workflow will fail because they were left behind on the VM and subsequent access or localization will fail. Checking for their existence when we evaluate the outputs allows to fail a bit earlier but it has a potential performance cost like we said.
I think we should consider dynamic outputs as a feature that is currently partially implemented (works for `cwl.output.json` but not for this tsv case in WDL as you caught it).

For example we could update the test and add a workflow output like

```wdl
String out = read_string(output_file_table.table[0][0]
```

which would probably fail the test although it's valid WDL.